### ISSUE TITLE
SHDR: EVENT MESSAGE handling

### ISSUE THREAD

**Issue by MRIIOT** [TrakHound/MTConnect.NET#19]
I don't believe Message is handled according to spec.

https://github.com/mtconnect/cppagent#adapter-agent-protocol-version-17

**Comment by MRIIOT** [comment]
I guess the adapter doesn't really know to treat it differently using `ShdrDataItem`.  Am I missing something?

**Comment by PatrickRitchie** [comment]
I don't believe I have support for that. I'm actually not entirely sure what the Message you referenced, and I'm assuming the Message class in the [dot_net_sdk](https://github.com/mtconnect/dot_net_sdk/blob/master/src/MTCDataItem.cs), is supposed to represent. In the source code comment it says it is for an Event but it has a Code property which I believe is for an Alarm DataEntity which has been deprecated and replaced with Condition. This seems to be the same format that is referenced in the SHDR protocol link you sent.

I do intend to support Alarms, in order to support legacy adapters/agents, but I just haven't implemented it yet (it's on my to do list).

My understanding of this might be wrong so I'll see if I can get some more information.

**Comment by MRIIOT** [comment]
The next special format is the Message. There is one additional field, native_code, which needs to be included:

2014-09-29T23:59:33.460470Z|message|CHG_INSRT|Change Inserts

**Comment by hulinning2** [comment]
By MTConnect spec, Message could be any string.  Only Condition has a format with native code.

**Comment by PatrickRitchie** [comment]
After getting some more information about this, I will add support for it even though it is not in the current MTConnect spec. This will make sure this library is compatible with the cpaagent and dot_net_sdk projects that many people already use.

I will add a new ShdrMessage class to handle the SHDR string and I will need to reorganize some of the ways I'm currently handling observations/streams to allow this which I will try to add to the next commit.

**Comment by hulinning2** [comment]
It is definitely in 1.3 spec.

![image](https://user-images.githubusercontent.com/4930742/156060536-a26f0b4f-2088-4ec4-a85a-8d571429ac82.png)

They misplaced it in other specs.

**Comment by PatrickRitchie** [comment]
@hulinning2  Yes that is my understanding that it was in a previous version. Although I still can't find where a NativeCode is specified for anything other than Condition, even in the previous versions.

**Comment by hulinning2** [comment]
> @hulinning2 Yes that is my understanding that it was in a previous version. Although I still can't find where a NativeCode is specified for anything other than Condition, even in the previous versions.

There is no native code in 1.3 spec for MESSAGE TAG. just a string.

**Comment by PatrickRitchie** [comment]
Based on this discussion and from gathering other information, each Type can technically add Attributes specific to that Type. This means that MESSAGE adds a NativeCode attribute (as you mentioned) along with AssetChanged and AssetRemoved both add a AssetType attribute.

This will need to be handled substantially differently than what was my initial approach. Since this can be dynamic and since I want this library to support custom types (not known at compile time), I won't be able to use the standard XML serializer since, to my knowledge, each class and each property must be defined at compile time.

I'm in the process of redesigning this serialization for XML and JSON and it looks like everything will work much better and support any combination of Category, Representation, and Type.

**Comment by PatrickRitchie** [comment]
This should be fixed in commit 0e50b3ba0c3403b4c86f5813944cf991ec523c38

Observations are now handled substantially differently behind the scenes and now support any property/value combination. Message should work similar to below:

Adapter code:
```c#
var adapter = new ShdrAdapter("demo");
adapter.Start();

var message = new ShdrMessage("message", "Change Inserts", "CHG_INSRT");
adapter.AddMessage(message);
```

SHDR Line:
> 2022-03-25T16:19:01.4900000Z|message|CHG_INSRT|Change Inserts

Agent Xml Output
```xml
<MTConnectStreams xmlns:m="urn:mtconnect.org:MTConnectStreams:1.8" xmlns="urn:mtconnect.org:MTConnectStreams:1.8" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="urn:mtconnect.org:MTConnectStreams:1.8 /schemas/MTConnectStreams_1.8.xsd">
    <Header instanceId="1648225092" version="0.3.1.0" sender="DESKTOP-HV74M4N" bufferSize="131072" firstSequence="1" lastSequence="116" nextSequence="117" creationTime="2022-03-25T16:19:19.2570774Z"/>
    <Streams>
        <DeviceStream name="demo" uuid="demo.01">
            <ComponentStream component="Path" componentId="path1" name="path">
                <Events>
                    <Message dataItemId="message" nativeCode="CHG_INSRT" sequence="115" timestamp="2022-03-25T16:19:01.4900000Z">Change Inserts</Message>
                </Events>
            </ComponentStream>
        </DeviceStream>
    </Streams>
</MTConnectStreams>
```

**Comment by MRIIOT** [comment]
Confirmed here.  Works with and without native code.
